relay-quotas/src/redis.rs

```diff
@@ -125,6 +137,189 @@ impl std::ops::Deref for RedisQuota<'_> {
     }
 }
 
+/// Possible outcomes from attemping to decrement a global counter budget.
+#[derive(Debug)]
+enum DecOutcome {
+    UpstreamExceeded,
+    NoRateLimit,
+    KeyMissing,
+    PoisonedLock,
+    SlotExpired,
+    BudgetEmpty,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum GlobalRateLimitError {
+    #[error("quota key missing")]
+    KeyMissing,
+    #[error("failed to access RWLock")]
+    PoisonedLock,
+    #[error("slot expired")]
+    SlotExpired,
+    #[error("budget empty")]
+    BudgetEmpty,
+    #[error("failed to communicate with redis")]
+    Redis,
+}
+
+/// Counters used as a cache for global quotas.
+///
+/// We might have different global counters for different conditions (e.g. different namespaces).
+/// We use an RwLock because the counter itself is an `AtomicUsize`. Atomics can be modified with
+/// a shared reference, meaning that we can have multiple threads updating the counter concurrently,
+/// as long as there's no writers.
+#[derive(Clone, Default)]
+pub struct GlobalCounters(Arc<RwLock<hashbrown::HashMap<BudgetKey, BudgetState>>>);
```

Comment:
Does this need to be pub? Its members seem to be private.

Reply:
oh, good call!